upup/pkg/kutil/rollingupdate_cluster_validate_drain.go

```diff
@@ -0,0 +1,313 @@
+
+	masterGroups := make(map[string]*CloudInstanceGroup)
+	nodeGroups := make(map[string]*CloudInstanceGroup)
+	bastionGroups := make(map[string]*CloudInstanceGroup)
+	for k, group := range groups {
+		switch group.InstanceGroup.Spec.Role {
+		case api.InstanceGroupRoleNode:
+			nodeGroups[k] = group
+		case api.InstanceGroupRoleMaster:
+			masterGroups[k] = group
+		case api.InstanceGroupRoleBastion:
+			bastionGroups[k] = group
+		default:
+			return fmt.Errorf("unknown group type for group: %q", group.InstanceGroup.ObjectMeta.Name)
+		}
+	}
+
+	// Upgrade bastions first; if these go down we can't see anything
+	{
+		var wg sync.WaitGroup
+
+		for k, bastionGroup := range bastionGroups {
+			wg.Add(1)
+			go func(k string, group *CloudInstanceGroup) {
+				resultsMutex.Lock()
+				results[k] = fmt.Errorf("function panic")
+				resultsMutex.Unlock()
+
+				defer wg.Done()
+
+				rollingUpdateData := c.CreateRollingUpdateData(instanceGroups, true)
+
+				err := group.RollingUpdateDrainValidate(rollingUpdateData)
+
+				resultsMutex.Lock()
+				results[k] = err
+				resultsMutex.Unlock()
+			}(k, bastionGroup)
+		}
+
+		wg.Wait()
+	}
+
+	// Upgrade master next
+	{
+		var wg sync.WaitGroup
+
+		// We run master nodes in series, even if they are in separate instance groups
+		// typically they will be in separate instance groups, so we can force the zones,
+		// and we don't want to roll all the masters at the same time.  See issue #284
+		wg.Add(1)
+
+		go func() {
+			for k := range masterGroups {
+				resultsMutex.Lock()
+				results[k] = fmt.Errorf("function panic")
+				resultsMutex.Unlock()
+			}
+
+			defer wg.Done()
+
+			for k, group := range masterGroups {
+				rollingUpdateData := c.CreateRollingUpdateData(instanceGroups, false)
+
+				err := group.RollingUpdateDrainValidate(rollingUpdateData)
+				resultsMutex.Lock()
+				results[k] = err
+				resultsMutex.Unlock()
+
+				// TODO: Bail on error?
+			}
+		}()
+
+		wg.Wait()
+	}
+
+	// Upgrade nodes, with greater parallelism
+	{
+		var wg sync.WaitGroup
+
+		for k, nodeGroup := range nodeGroups {
+			wg.Add(1)
+			go func(k string, group *CloudInstanceGroup) {
+				resultsMutex.Lock()
+				results[k] = fmt.Errorf("function panic")
+				resultsMutex.Unlock()
+
+				defer wg.Done()
+
+				rollingUpdateData := c.CreateRollingUpdateData(instanceGroups, false)
+
+				err := group.RollingUpdateDrainValidate(rollingUpdateData)
+
+				resultsMutex.Lock()
+				results[k] = err
+				resultsMutex.Unlock()
+			}(k, nodeGroup)
+		}
+
+		wg.Wait()
+	}
+
+	for _, err := range results {
+		if err != nil {
+			return err
+		}
+	}
+
+	glog.Infof("\nRolling update completed!\n")
+	return nil
+}
+
+// CreateRollingUpdateData creates a RollingUpdateClusterDrainValidate struct.
+func (c *RollingUpdateClusterDrainValidate) CreateRollingUpdateData(instanceGroups *api.InstanceGroupList, isBastion bool) *RollingUpdateDataDrainValidate {
+	return &RollingUpdateDataDrainValidate{
+		Cloud:             c.Cloud,
```

Comment:
Also, it did not start out that way ... kinda ended up in that direction.  Same question about the two refactors you asked for.  Can we push them into another PR?  Completely fair for you to say no ...